pkg/memory/short_term.go

```diff
@@ -0,0 +1,157 @@
+// that have already been performed.
+
+package memory
+
+import (
+	"reflect"
+	"sync"
+	"time"
+)
+
+// ShortTermMemoryBuilder builds of short term memory objects.
+//
+type ShortTermMemoryBuilder struct {
+	// How long to remember actions.
+	duration time.Duration
+}
+
+// ShortTermMemory stores a set of items for a given period of time.
+//
+type ShortTermMemory struct {
+	// How long to remember actions.
+	duration time.Duration
+
+	// There will be a cell for each action stored, containing the action itself and the time it was
+	// added to the memory.
+	cells []*ShortTermCell
+
+	// Mutex used to prevent simultaneous updates of the data structures.
+	mutex *sync.Mutex
+}
+
+// ShortTermCell stores each individual action, and the time it was added to the memory.
+//
+type ShortTermCell struct {
+	// The item stored in the cell.
+	item interface{}
+
+	// The time when the cell was created or updated.
+	stamp time.Time
+}
+
+// NewShortTermMemoryBuilder creates a builder that can create short term memory objects.
+//
+func NewShortTermMemoryBuilder() *ShortTermMemoryBuilder {
+	b := new(ShortTermMemoryBuilder)
+	return b
+}
+
+// Duration sets how long objects in the memory will be remembered. The default is zero, which means
+// that objects won't be remembered at all.
+//
+func (b *ShortTermMemoryBuilder) Duration(duration time.Duration) *ShortTermMemoryBuilder {
+	b.duration = duration
+	return b
+}
+
+// Build creates a new short term memory object with the configuration stored in the builder.
+//
+func (b *ShortTermMemoryBuilder) Build() (m *ShortTermMemory, err error) {
+	m = new(ShortTermMemory)
+	m.duration = b.duration
+	m.cells = make([]*ShortTermCell, 0)
+	m.mutex = &sync.Mutex{}
+	return
+}
+
+// Add adds a new item to the memory.
+//
+func (m *ShortTermMemory) Add(item interface{}) {
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+
+	cell := m.findMatchingCell(item)
+	if cell == nil {
+		cell = new(ShortTermCell)
+		cell.item = item
+		m.cells = append(m.cells, cell)
+	}
+	cell.stamp = time.Now()
+}
+
+func (m *ShortTermMemory) Has(item interface{}) bool {
+	// Purge cells before checking.
+	m.CleanShortTermMemory()
+
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+
+	return m.findMatchingCell(item) != nil
+}
+
+// Len returns the number of items inside the memory.
+//
+func (m *ShortTermMemory) Len() int {
+	m.CleanShortTermMemory()
+
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+
+	return len(m.cells)
+}
+
+// PurgeExpiredCells finds the aged cells and removes them.
+//
+func (m *ShortTermMemory) purgeExpiredCells() {
+	now := time.Now()
+	for idx, cell := range m.cells {
+		age := now.Sub(cell.stamp)
+		// purge cells in place.
+		if age >= m.duration {
+			// zeroing the value of the cell so that it wouldn't be refrenced by the underlying array
+			// causing GO's grabage collector to collect the allocated memory.
+			cell = nil
+			m.cells = append(m.cells[:idx], m.cells[idx+1:]...)
+		}
```

Comment:
It seems the timestamps in `cells` are monotonously increasing
That means that you could break when age < duration

Reply:
you mean do this?
```
if age < m.duration {
  m.cells = m.cells[idx:]
}
```
Yeah this is true. But I like this implementation and since we don't *add* more cells but rather substract cells than the capacity of the underlying array is less or equal to the amount of cells we add in the `append` thus the complexity is **amortized constant** (see [here](https://groups.google.com/forum/#!msg/golang-nuts/o5rFNqd0VjA/HzJHwgl1y6MJ)).
Plus if you don't erase the cells (i.e `cell = nil`) the underlying struct remains the same size...

Reply:
This is not true you have 'nil out' all the cells before you call append otherwise memory explodes and you don't really do append meaning it is O(retired items) which can be O(len). You assume that the interval is small but it doesn't have to be..

Reply:
Oh you're first idea should work just fine! I thought you meant what I wrote..